Index: firewall4-2024.12.18~18fc0ead/root/usr/share/ucode/fw4.uc
===================================================================
--- firewall4-2024.12.18~18fc0ead.orig/root/usr/share/ucode/fw4.uc
+++ firewall4-2024.12.18~18fc0ead/root/usr/share/ucode/fw4.uc
@@ -3147,6 +3147,7 @@ return {
 			switch (proto.name) {
 			case "tcp":
 			case "udp":
+			case "icmp":
 				sport = snat.src_port;
 				dport = snat.dest_port;
 				rport = snat.snat_port;
